cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -293,18 +303,42 @@ public RollbackDetail publish(StoreRequest request)
     if (!batch.isEmpty()) {
       int sequenceId = 0;
       List<Mutation> batchCopy = new ArrayList<>(batch.size());
+      long estimatedBatchSize = 0;
       // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
       // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
       // not publish_ts > last_message.publish_ts
       while (!batch.isEmpty()) {
         StoreRequest headRequest = batch.poll();
         for (byte[] payload : headRequest) {
-          //TODO: [CDAP-21094] Breakdown messages with larger payload into parts.
-          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
-              .set(SEQUENCE_ID_FIELD).to(sequenceId++).set(PAYLOAD_SEQUENCE_ID_FIELD).to(0)
-              .set(PUBLISH_TS_FIELD).to("spanner.commit_timestamp()").set(PAYLOAD_FIELD)
-              .to(ByteArray.copyFrom(payload)).build();
-          batchCopy.add(mutation);
+          int payloadSequenceId = 0;
+          List<Mutation> mutationsForCurrentPayload = new ArrayList<>();
+          int remainingParts = (payload.length - 1) / MAX_PAYLOAD_SIZE_IN_BYTES;
+
+          while (payload.length > MAX_PAYLOAD_SIZE_IN_BYTES) {
+            // Split payload if it exceeds the maximum size allowed by Spanner (10 MB).
+            // We are keeping max size as 9MB so that we do not touch the spanner threshold.
+            byte[] payloadSegment = Arrays.copyOfRange(payload, 0, MAX_PAYLOAD_SIZE_IN_BYTES);
+            Mutation mutation = constructPublishMutation(headRequest.getTopicId(), sequenceId,
```

Comment:
use consistent naming. E.g., payloadSegment and remainingParts are related, but their naming are not consistent. 
e.g., rename to payloadChunk and remainingChunksNumber